### PR TITLE
[Caffe2] Enable MIOpen Conv3D Operator.

### PR DESCRIPTION
**Pull request by lcskrishna**
1. Enabled conv3d support.
2. Enabled a test case for the same.

### CODE REVIEW

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 1]
```diff
@@ -583,6 +689,7 @@ bool MIOPENConvGradientOp::DoRunWithType() {
       bwdWeiAlgo_ = perf.bwd_weights_algo;
     }
   }
+
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: superfluous white space change

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 2]
```diff
@@ -667,4 +774,9 @@ bool MIOPENConvGradientOp::RunOnDevice() {
 
 REGISTER_MIOPEN_OPERATOR(Conv, MIOPENConvOp);
 REGISTER_MIOPEN_OPERATOR(ConvGradient, MIOPENConvGradientOp);
+
+// 3-D Convolutions.
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: just remove comment, I think the naming is clear enough, remove empty line before as well

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 3]
```diff
@@ -252,48 +253,101 @@ bool MIOPENConvOp::DoRunWithType() {
...
       mio_input_dims_ = X.sizes().vec();
-      MIOPEN_ENFORCE(miopenSet4dTensorDescriptor(
-          bottom_desc_, miopenTypeWrapper<T_X>::type, N, C, H, W));
+      if (mio_input_dims_.size() == 4) {
+        MIOPEN_ENFORCE(miopenSet4dTensorDescriptor(
+            bottom_desc_, miopenTypeWrapper<T_X>::type, N, C, H, W));
+      } else {
+        vector<int> input_dims (begin(mio_input_dims_), end(mio_input_dims_));
+        MIOPEN_ENFORCE(miopenSetTensorDescriptor(
+            bottom_desc_, miopenTypeWrapper<T_X>::type, input_dims.size(), input_dims.data(), nullptr));
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: is nullptr allowed here? MIOpen doc is silent on that.

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 4]
```diff
@@ -252,48 +253,101 @@ bool MIOPENConvOp::DoRunWithType() {
...
+            kernel_h(),
+            kernel_w()));
+      } else {
+        vector<int> weight_dims (begin(mio_weight_dims_), end(mio_weight_dims_));
+        MIOPEN_ENFORCE(miopenSetTensorDescriptor(
+            weight_desc_,
+            miopenTypeWrapper<T_W>::type,
+            mio_weight_dims_.size(),
+            weight_dims.data(),
+            nullptr));
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: same question here.

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 5]
```diff
@@ -252,48 +253,101 @@ bool MIOPENConvOp::DoRunWithType() {
...
-        bottom_desc_,
-        weight_desc_,
-        &N_out,
-        &C_out,
-        &H_out,
-        &W_out));
+    
+    mio_output_dims_ = Y->sizes().vec();
+    int output_dim_size = mio_input_dims_.size(); 
+    vector<int> output_dims (begin(mio_output_dims_), end(mio_output_dims_));
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: is this required to be done on every invocation?

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 6]
```diff
@@ -252,48 +253,101 @@ bool MIOPENConvOp::DoRunWithType() {
...
+    } else {
+      MIOPEN_ENFORCE(miopenSetTensorDescriptor(
+          top_desc_, miopenTypeWrapper<T_X>::type, output_dims.size(), output_dims.data(), nullptr));
+    }
 
     if (InputSize() == 3) {
-      MIOPEN_ENFORCE(miopenSet4dTensorDescriptor(
-          bias_desc_, miopenTypeWrapper<T_B>::type, 1, M, 1, 1));
+      if (mio_input_dims_.size() == 4) {
+        MIOPEN_ENFORCE(miopenSet4dTensorDescriptor(
```
Comment (iotamudelta) [inline caffe2/operators/hip/conv_op_miopen.hip]: General question: could we not use the generic miopenSetTensorDescriptor also for the 4D case and eschew these if/else here?

File: caffe2/operators/hip/conv_op_miopen.hip (hunk 3 above)
Comment (lcskrishna) [reply caffe2/operators/hip/conv_op_miopen.hip]: Yes, confirmed with @ce1adon before implementation.

File: caffe2/operators/hip/conv_op_miopen.hip (hunk 6 above)
Comment (lcskrishna) [reply caffe2/operators/hip/conv_op_miopen.hip]: Haven't considered about refactoring the existing code, but yes 4D case can also be part of miopenSetTensorDescriptor. Let me clean up the code.

File: caffe2/operators/hip/conv_op_miopen.hip (hunk 1 above)
Comment (lcskrishna) [reply caffe2/operators/hip/conv_op_miopen.hip]: Done

File: caffe2/operators/hip/conv_op_miopen.hip (hunk 2 above)
Comment (lcskrishna) [reply caffe2/operators/hip/conv_op_miopen.hip]: Done

File: caffe2/operators/hip/conv_op_miopen.hip (hunk 5 above)
Comment (lcskrishna) [reply caffe2/operators/hip/conv_op_miopen.hip]: No, it is not done on every invocation. 
This is done only when the if (input_changed ) is true in Line number 252 and 470.

File: caffe2/operators/hip/conv_op_miopen.hip [hunk 7]
```diff
@@ -411,7 +414,7 @@ bool MIOPENConvGradientOp::DoRunWithType() {
 
   CAFFE_ENFORCE(X.ndim() >= 3 && X.ndim() <= 5);
   CAFFE_ENFORCE(
-      Weight.ndim() == 4,
+      Weight.ndim() == 4 || Weight.ndim() == 5,
       "ConvGradient op with MIOpen engine is supported only for 2D convolutions");
```
Comment (ashishfarmer) [inline caffe2/operators/hip/conv_op_miopen.hip]: nit: The enforce message should say "2D and  3D convolutions"

**Comment by ashishfarmer** [comment]
Once the comment is fixed with correct message, it looks good to be merged

**Pull request merged**